### PR TITLE
Referrer exclusion segmentation

### LINKED ISSUE #386: Segmentation By Referrer Suppression
In addition to the existing referrer logic, it would be useful to create a segment that does _not_ show to users who have come in from a particular domain, i.e. a pop-up that would be shown to everyone except traffic from google.com.

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #386 

### How to test the changes in this Pull Request:

Follow testing steps in https://github.com/Automattic/newspack-plugin/pull/831

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: api/campaigns/class-campaign-data-utils.php [hunk 1]
```diff
@@ -27,6 +27,23 @@ public static function is_donor( $client_data ) {
...
+	/**
+	 * Compare page referrer to a list of referrers.
+	 *
+	 * @param string $page_referrer_url Referrer to compare to.
+	 * @param string $referrers_list_string Comma-separated referrer domains list.
+	 */
+	public static function does_referrer_match( $page_referrer_url, $referrers_list_string ) {
+		$referer_domain = parse_url( $page_referrer_url, PHP_URL_HOST ); // phpcs:ignore WordPress.WP.AlternativeFunctions.parse_url_parse_url, WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		// Handle the 'www' prefix – assume `www.example.com` and `example.com` referrers are the same.
+		$referer_domain_alternative = strpos( $referer_domain, 'www.' ) === 0 ? substr( $referer_domain, 4 ) : "www.$referer_domain";
```
Comment (dkoo) [inline api/campaigns/class-campaign-data-utils.php:39]: This made me realize that any subdomain other than `www` won't be considered a match. e.g. if I set my segment to exclude `google.com`, and the referrer is `https://gmail.google.com`, it won't be considered a match and the campaign will still be displayed. Also true of the regular referrer option. Should we clarify that URLs with subdomains must be specified in the field if they want them to match the segment?

**Review by dkoo — CHANGES_REQUESTED** [review]

File: api/campaigns/class-campaign-data-utils.php (hunk 1 above)
Comment (adekbadek) [reply api/campaigns/class-campaign-data-utils.php:39]: Now I wonder if we should maybe disregard the subdomain - just check the domain. Referral from `images.example.com` is still a referral from `example.com`, right?

Comment (dkoo) [reply api/campaigns/class-campaign-data-utils.php:39]: Yeah, that's what I was wondering. My gut says we should just look at the root domain and ignore subdomains.

Comment (adekbadek) [reply api/campaigns/class-campaign-data-utils.php]: Right, done in d16574b

**Review by dkoo — APPROVED** [review]
👍

**Pull request merged**